### ISSUE TITLE
HtmlToWmlConverter removes pre, tt, code, kbd, samp etc.

### ISSUE THREAD

**Issue by lajjne** [OfficeDev/Open-Xml-PowerTools#19]
All text inside `<pre>`, `<tt>`, `<code>`, `<kbd>` and `<samp>` is removed when converting to docx with HtmlToWmlConverter

**Comment by EricWhiteDev** [comment]
@lajjne Very good point.  This needs addressed.  I have assigned to myself.

**Comment by lajjne** [comment]
I found that this is a general problem with all elements that is not explicitly handled in the Transform method.

**Comment by EricWhiteDev** [comment]
Do you have a suggestion about how to handle all elements that are not explicitly handled?

In the case of  `<pre>`, `<tt>`, `<code>`, `<kbd>` and `<samp>`, they should be converted to a fixed font, I believe.

**Comment by lajjne** [comment]
I have a couple of ideas. I'll try to submit a pull request, but since I don't fully understand the code base you will probably need to tweak it a bit to fit in with the rest of the code :)

**Comment by EricWhiteDev** [comment]
I'd love to see the PR, but in any case, don't hassle too much - these changes are pretty easy - can knock them out in an hour or two, and I have some specific ideas of how to implement them.

In general, I'm thinking that for pre, tt, code, kbd, and samp, I will define a linked paragraph/character style with a fixed width font.  It is possible to also put the font / font size info into the HtmlToWmlConverterSettings class, which will enable developers to tweak to their heart's desire.

**Comment by lajjne** [comment]
I'm not sure you need to do anything special for `tt`, `code`, `kbd`, and `samp`. The existing defaultCss already handles the styling. All I did to get it working was to add this in the Transform method:

````
if (element.Name == "tt" || element.Name == "code" || element.Name == "kbd" || element.Name == "samp") {
    return element.Nodes().Select(n => Transform(n, settings, wDoc, nextExpected));
}
````

Handling `pre` was a little more difficult since it needs to keep the existing whitespace and line breaks. I ended up with something like this but you can probably come up with a more elegant solution.

````
if (element.Name == "pre") {
    // TODO: element.Value removes any <br /> inside the <pre> which is not what we want
    var pre = element.Value;
    // get the individual lines inside the <pre> element
    var lines = pre.Split(new char[] { '\n' }, StringSplitOptions.None);
    XElement rpr = GetRunProperties(element, settings);
    var r = new XElement(W.r, rpr);
    foreach (var line in lines) {
        r.Add(new XElement(W.t, new XAttribute(XNamespace.Xml + "space", "preserve"), line));
        r.Add(new XElement(W.br));
    }
    var p = new XElement(W.p, r);
    return p;
}
````

I've also been thinking about how to handle unknown (or undefined) elements in the transform. First I think you should add all known HTML elements (see for example https://developer.mozilla.org/en/docs/Web/HTML/Element) to OpenXmlPowerTools.Xhtml and OpenXmlPowerTools.XhtmlNoNamespace and of course handle them in the Transform method. 

Any unknown elements should probably be treated like a `div` or similar so that they are not removed in the transform. See this section in the HTML spec. (http://www.w3.org/TR/html5/infrastructure.html#extensibility-0)

> User agents must treat elements and attributes that they do not understand as semantically neutral; leaving them in the DOM (for DOM processors), and styling them according to CSS (for CSS processors), but not inferring any meaning from them.

**Comment by EricWhiteDev** [comment]
@lajjne Ok, take a look now.  I've just pushed a fix for pre, code, samp, kbd, and tt.  Take a look at T1650.html and T1660.html.

I'll look at the hyperlinks issue next.

**Comment by EricWhiteDev** [comment]
Also, I'm leaving this open, as I want to go through the https://developer.mozilla.org/en/docs/Web/HTML/Element and http://www.w3.org/TR/html5/infrastructure.html#extensibility-0

**Comment by lajjne** [comment]
Works perfectly. Thanks. Also good that you are keeping it open until all other elements (even unknown) are handled. Just today I ran into a document where a `blockquote` was removed in the conversion.

**Comment by EricWhiteDev** [comment]
I want this module to be top notch - and it is fun to tie up these loose ends.  :-)  Technically, I am on vacation, so progress will happen as chances occur.

**Comment by EricWhiteDev** [comment]
@lajjne I've modified HtmlToWmlConverter to simply recursively process all descendants for unknown elements.  I added tests for section, article elements, and it works properly.  I am sure that there is further work that we eventually want to do to this module in this area to modify the rendering of additional elements, but there are other high-pri objectives that we want to take care of first.

This change should, in general, make this module much more robust for varieties of HTML files that I haven't considered.
